### PR TITLE
Only permit scalars and weighted sums for `assign`

### PR DESCRIPTION
**Pull request by connorjward**
Associated issue: #2556

We also need to merge changes in pyadjoint: https://github.com/dolfin-adjoint/pyadjoint/pull/93.

### CODE REVIEW

File: firedrake/assign.py
Comment (wence-) [inline firedrake/assign.py:22]: Don't hand traverse, instead write:
```
def sum(self, o, *operands):
    pass
```

Hand-traversal destroys the DAG-based visitor pattern with the result that you will run slower than you want.

File: firedrake/assign.py
Comment (wence-) [inline firedrake/assign.py:35]: Prefer to return values and glue them together coming back up the traversal rather than side-effectful modification of the visitor. That way you will be able to get away with only constructing a single `WeightedCoefficientCollector` (rather than one for every expression).

File: firedrake/assign.py
Comment (wence-) [inline firedrake/assign.py:38]: Does this mean that `f.assign(as_tensor([[1, 2], [3, 4]]) @ as_vector([a, b]))` will be rejected? That is a linear combination so should be safe.

File: firedrake/assign.py
Comment (wence-) [inline firedrake/assign.py:83]: This caching is unsafe if someone ever holds onto an `Assigner` object and runs it in a loop where the rvalue coefficients are modified elsewise.

File: firedrake/assign.py
Comment (wence-) [inline firedrake/assign.py:116]: Careful, this is unsafe in the case that the lvalue or rvalues are mixed dats (the `_vec` is not a view into the data in that case)

File: firedrake/assign.py
Comment (connorjward) [reply firedrake/assign.py:38]: Running this gives me an `unsupported operand type(s) for @` error. Are you sure that UFL supports this sort of thing?

Comment (wence-) [reply firedrake/assign.py:38]: Ah ok, I see.

**Comment by connorjward** [comment]
A quick experiment shows that this can speed things up quite a bit. I've been running the [DG advection](https://firedrakeproject.org/demos/DG_advection.py.html) demo on my branch vs `master`. Below are the flamegraphs I get on my laptop (with a hot cache). The first shows current `master` and the bottom one is my code. I found the code took ~25s on `master` and mine took ~20s. You can see that the `assign` bar is much smaller in the bottom one.

old1
new1

I have a few comments:
- We still need to apply a tree visitor to the expression. Caching on a persistent expression would let us avoid this.
- Alternatively, using a persistent `Assigner` object makes pretty much all overhead go away. The flamegraph below, where I tried this, `assign` takes 0.2s out of a total 18s.

persistent_assigner

File: firedrake/assemble.py
Comment (JDBetteridge) [inline firedrake/assemble.py:375]: This FIXME probably needs addressing before merge

File: firedrake/assign.py
Comment (JDBetteridge) [inline firedrake/assign.py:117]: Fix this?

File: firedrake/function.py
Comment (JDBetteridge) [inline firedrake/function.py:386]: Is there a reason we can't `from firedrake import assign` at the top and `assign.Assigner.assign()`? Or just import the things we need from `assign.py`

File: firedrake/utils.py
Comment (JDBetteridge) [inline firedrake/utils.py:100]: This can be achieved more efficiently using `itertools`, see the recipe for `partition`, something like:
```python
import itertools
def itersplit(pred, iterable):
    t1, t2 = itertools.tee(iterable)
    return tuple(filter(pred, t1)), tuple(itertools.filterfalse(pred, t2))
```

Shamelessly stolen from: https://docs.python.org/3/library/itertools.html#itertools-recipes

**Review by JDBetteridge — CHANGES_REQUESTED** [review]
I think we are mainly just waiting on some things to be fixed for adjoint

File: .github/workflows/build.yml
Comment (connorjward) [reply .github/workflows/build.yml:45]: We are.

File: firedrake/function.py
Comment (connorjward) [reply firedrake/function.py:386]: Because `firedrake/assign.py` has `from firedrake.function import Function` at the top (import loop).

File: firedrake/assemble.py
Comment (connorjward) [reply firedrake/assemble.py:375]: This is essentially just a documentation fix to go with https://github.com/firedrakeproject/firedrake/issues/2581. I thought it would do no harm to merge alongside this PR.

File: firedrake/assign.py
Comment (connorjward) [reply firedrake/assign.py:117]: I argue that this is a separate issue to resolve. Current adjoint approaches will still work fine. I've just added an extra entry point for `assign` that lacks the functionality.

**Review by dham — CHANGES_REQUESTED** [review]

File: firedrake/utils.py
Comment (connorjward) [reply firedrake/utils.py:100]: I'd like to claim that my solution works fine and so this isn't a big deal.

**Review by JDBetteridge — DISMISSED** [review]

**Review by dham — DISMISSED** [review]

**Review by ReubenHill — CHANGES_REQUESTED** [review]
Please check that docs are up to date

**Comment by ReubenHill** [comment]
> Please check that docs are up to date

Poke me to approve when done

**Comment by connorjward** [comment]
@ReubenHill I think the docs all look fine. This is ready to go.

File: tests/vertexonly/test_poisson_inverse_conductivity.py
Comment (ReubenHill) [inline tests/vertexonly/test_poisson_inverse_conductivity.py:32]: Quick query before I approve, why was this changed?

Comment (connorjward) [reply tests/vertexonly/test_poisson_inverse_conductivity.py:32]: With [this line](https://github.com/firedrakeproject/firedrake/blob/master/tests/vertexonly/test_poisson_inverse_conductivity.py#L18) above it in the file we were effectively switching on the adjoint twice. Since we use a [counter](https://github.com/dolfin-adjoint/pyadjoint/blob/master/pyadjoint/tape.py#L31) to track whether or not the adjoint is enabled this meant that switching the adjoint off with `with stop_annotating():` only incremented the counter once and we still thought that we were supposed to be taping.

This was causing tests to fail because I now `raise NotImplementedError` if someone is inside either `FormAssembler.assemble` or `Assigner.assign` and taping is switched on.

Comment (ReubenHill) [reply tests/vertexonly/test_poisson_inverse_conductivity.py:32]: But there is no `with stop_annotating():` here? Or am I being daft?

Comment (connorjward) [reply tests/vertexonly/test_poisson_inverse_conductivity.py:32]: We call `with stop_annotating():` inside the adjoint (e.g. [here](https://github.com/firedrakeproject/firedrake/blob/master/firedrake/adjoint/assembly.py#L18)). I think the idea is to make sure that we aren't adding blocks to the tape that are 'inside' other blocks. An example would be that we want to tape a `solve` block but not the `assemble` calls that happen inside the `solve`.

Comment (ReubenHill) [reply tests/vertexonly/test_poisson_inverse_conductivity.py:32]: Okay I think I follow. Either way it's nice to see that this "just in case" code can be got rid of! I think previously it wasn't clear in either case where we needed it so it's good that now you got an error thrown

**Review by ReubenHill — APPROVED** [review]

**Pull request merged**
